corehq/apps/enterprise/models.py

```diff
@@ -63,3 +71,19 @@ class EnterpriseMobileWorkerSettings(models.Model):
     enable_auto_deactivation = models.BooleanField(default=False)
     inactivity_period = models.IntegerField(default=90)
     allow_custom_deactivation = models.BooleanField(default=False)
+
+    def deactivate_mobile_workers_by_inactivity(self, domain):
+        user_query = UserES().domain(domain).mobile_users().is_active()
+        user_data = user_query.source([
+            '_id',
+            'created_on',
+        ]).run()
+        for user in user_data.hits:
+            inactive_days = get_days_since_last_form_submission_for_user_id(
+                domain, user['_id']
```

Comment:
It seems inefficient to do a separate query for every user in a domain. Is there any possibility of more efficiently finding the users that should be deactivated? Or maybe possible to get most recent submission dates for a batch of users in a single query?

Reply:
i'm trying to figure out how I could do such a query but also catch the users who have made no submissions at all since their `created_on` date. any suggestions of whether we have something that does a similar lookup?

Reply:
I don't know if we have SQL indexes in the shard DBs that would allow us to efficiently query forms by `domain`, `user_id`, and `received_on`? Something like
```sql
SELECT MAX(received_on)
FROM forms
WHERE domain = :domain
    AND user_id IN (...) -- list of user ids
GROUP BY domain, user_id
```
That query would need to run against each shard, and then the results could be combined. The shard queries could be run in parallel if PL/Proxy functions were added for the shard query as well as proxy aggregation of shard results.

Or maybe there is a way to do a query like that on Elasticsearch?

Reply:
oh that is a good point...this data is now in SQL, at least most of it. the issue being `created_on` is still in couch...so I'm still tripped up on that one.

Reply:
@millerdev so I reviewed the scale of the projects that will be using this feature, and the most users they will start out with will be under 10k in total (3k on average) within the next 2 years. Therefore, the number of auto-deactivations per month should be in the low 1000s at most if not in the 100s. 

With this in mind, I'm not sure optimizing this code prior to migrating the `CommCareUser` models to SQL is the best use of time? What I can do is set up a DataDog metric in the task to track the run times, and have it flag when it gets close to something that we should be concerned about (maybe over 20 minutes?). Thoughts?

I think once the couch to SQL is done for `CommCareUser`s then this bit can be revisited.

Reply:
Is this what you're looking for?
https://github.com/dimagi/commcare-hq/blob/567e48f97c0cb0130f4a03e4a64bdf73c981f9f3/corehq/apps/reports/analytics/esaccessors.py#L41-L67
Not sure that does exactly what you're talking about, but it looks useful

Reply:
thanks @esoergel I had seen that but was still stuck on the `created_on` filter. I had a solution in mind to work around this, but I think I found an even better approach. Thanks Ben R for creating these handy reporting metrics in the user model. ^_^

@millerdev yes, definitely understand, and great points. I think adding a metric to the task is good to also gauge that it's running properly.

However, I am still struggling to find a workaround on bulk deactivating `CommCareUser`s one at a time. The model that lives in couch still needs to be updated, and we have to create audit trails for each user change. Please see my other comment and I appreciate any suggestions.

Reply:
> struggling to find a workaround on bulk deactivating `CommCareUser`s one at a time. The model that lives in couch still needs to be updated

If the Couch docs still need to be updated then we have no choice but to update each one individually, I think. Couch can't do bulk updates, right?. Possibly could do that after bulk-updating SQL models to at least reduce the number of SQL queries.

Another idea I just had would be to introduce a delay between each iteration, which would make the overall task take longer, but would spread out the load. We could keep this in mind as something to think about if this task causes performance issues in other places.

Reply:
Couch can definitely do bulk updates:
```python
CouchUser.get_db().bulk_save(users)
```
`IterDB` or `iter_update` will handle chunking for you too.  The problem is, the `save` methods won't get called, so any signals or other things we've added there won't be fired, like the auditing Biyeun mentioned.  Pillows should be good though (I'm 80% sure), since the changes feed _will_ have bulk updated docs.

Reply:
alright! made some changes... @millerdev and I offlined. I feel the solution now is about as good as it gets until we migrate `CommCareUsers` off of couch db :)